### PR TITLE
Impl arithmetic operations and polynomial interpolation over Galois fields GF(2^192)

### LINKED ISSUE #473: [800 SigmaUSD] Arithmetic operations and polynomial interpolation over Galois fields GF(2^192)
### Implementation details
Extract https://github.com/ergoplatform/sigma-rust/blob/develop/ergotree-interpreter/src/sigma_protocol/gf2_192.rs and https://github.com/ergoplatform/sigma-rust/blob/develop/ergotree-interpreter/src/sigma_protocol/gf2_192poly.rs into a separate crate `gf2_192` and implement according to existing implementation in Scala https://github.com/ScorexFoundation/sigmastate-interpreter/blob/prepare-v5.0-evaluator/sigmastate/src/main/java/gf2t/GF2_192.java , https://github.com/ScorexFoundation/sigmastate-interpreter/blob/prepare-v5.0-evaluator/sigmastate/src/main/java/gf2t/GF2_192_Poly.java

### Testing 
Implement tests according to the tests in Scala version in https://github.com/ScorexFoundation/sigmastate-interpreter/blob/prepare-v5.0-evaluator/sigmastate/src/test/java/gf2t/GF2_192Test.java

### References
https://github.com/reyzin/GF2t

Existing Rust implementation for Galois fields GF(2^256):
https://github.com/djsweet/galois_2p8
https://github.com/geky/gf256

### PR DESCRIPTION
**Pull request by kettlebell**
Closes #473

### CODE REVIEW

File: gf2_192/src/gf2_192poly.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        // Try with arrays of length 0
+
+        // Constant polynomial with value 0
+        let mut p = Gf2_192Poly::interpolate(&[], &[], Gf2_192::new()).unwrap();
+        assert!(p.evaluate(0).is_zero());
+        assert!(p.evaluate(5).is_zero());
+
+        // Constant polynomial with value 17
+        let val_17 = Gf2_192::from(17);
+        p = Gf2_192Poly::interpolate(&[], &[], val_17).unwrap();
```
Comment (kettlebell) [inline gf2_192/src/gf2_192poly.rs:205]: Note that `sigmastate-interpreter` doesn't actually use empty arrays for these tests: https://github.com/ScorexFoundation/sigmastate-interpreter/blob/262a14d17ae3090fa97d1de1effd0b338562e4db/sigmastate/src/test/java/gf2t/GF2_192Test.java#L595-L599. Reyzin has a comment '0 is in the points array and valueAt0!=null, behavior is undefined', but these tests are counter-examples : ) 

Also the test on line 599 there is actually wrong! `p = GF2_192_Poly.interpolate(new byte[0], new GF2_192[0], val17);`. We can't have a polynomial taking values of 0 and 17 at zero! Strangely I assume the test is passing there, but it failed on my implementation.

File: gf2_192/src/gf2_192poly.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+}
+
+impl Gf2_192Poly {
+    /// Create the unique lowest-degree interpolating polynomial that passes through
+    /// `(0, value_at_zero)` and `(points[i], values[i])` for all `i = 0, ..(points.len() - 1)`.
+    /// Assumptions:
+    ///  - Elements of `points` must be distinct and must not contain `0`.
+    ///  - `points.len() == values.len()`. Note that `points` and `values` can be empty, resulting
+    ///    in a constant polynomial with value `value_at_zero`.
+    pub fn interpolate(
```
Comment (kettlebell) [inline gf2_192/src/gf2_192poly.rs:63]: I implemented a simplified form of this method from `sigmastate-interpreter` where `value_at_zero` must be present. This is in line with your indicated usage in `ergotree-interpreter`.

File: gf2_192/src/gf2_192poly.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            _t = p.coefficients[i] * r;
+            self.coefficients[i] = self.coefficients[i] + _t;
+        }
+        self.degree = p.degree;
+        self.coefficients[self.degree] = r;
+    }
+
+    /// Multiply `self` by `x + r`. Assumes that `self` is monic (i.e.
+    /// `self.coefficients[self.degree] == 1`)
+    fn multiply_by_linear_binomial(&mut self, r: i8) {
```
Comment (kettlebell) [inline gf2_192/src/gf2_192poly.rs:145]: This method is named `monicTimesMonomial` by Reyzin, but I checked the definitions and I'm sure it's not a monomial but rather a linear binomial.

File: gf2_192/src/gf2_192poly.rs (hunk 1 above)
Comment (greenhat) [reply gf2_192/src/gf2_192poly.rs:206]: Good catch! I see that both implementation and tests are written by Reyzin. :) Maybe @kushti you can shed some light here?

File: gf2_192/src/gf2_192poly.rs [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+        coefficients[0] = Gf2_192::from(constant_term);
+        let degree = 0;
+        Gf2_192Poly {
+            degree,
+            coefficients,
+        }
+    }
+}
+
+impl<'a> TryFrom<CoefficientsByteRepr<'a>> for Gf2_192Poly {
```
Comment (greenhat) [inline gf2_192/src/gf2_192poly.rs:170]: Could you pull in here the explanation for this from 
https://github.com/ScorexFoundation/sigmastate-interpreter/blob/262a14d17ae3090fa97d1de1effd0b338562e4db/sigmastate/src/main/java/gf2t/GF2_192_Poly.java#L40-L48
and/or add it to the `CoefficientByteRepr` type.

File: gf2_192/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+use gf2_192poly::Gf2_192PolyError;
+use thiserror::Error;
+
+pub mod gf2_192;
+pub mod gf2_192poly;
+
+/// Logical right shift of i64 values which is equivalent to the `>>>` operator in java/scala. Need
+/// this because rust does 'arithmetical shift right' on signed integers.
+/// See:
+///  - https://en.wikipedia.org/wiki/Logical_shift
```
Comment (greenhat) [inline gf2_192/src/lib.rs:32]: Could you please wrap URL in `<>` to make intra-doc linter happy? It's for properly rendering link on docs.rs.

**Review by greenhat — CHANGES_REQUESTED** [review]
This looks awesome! Great job! I left some comments. 
I won't pretend I understand how it works :) so I carefully compared it to the existing Java implementation.

@kushti Could you take a look at the raised question in the comments? I'd appreciate it if you take a look at the code as well.

**Comment by kettlebell** [comment]
> This looks awesome! Great job! I left some comments. I won't pretend I understand how it works :) so I carefully compared it to the existing Java implementation.

Thanks! I know a little of how it all works, but I did largely the same as you :)

File: gf2_192/src/gf2_192poly.rs (hunk 4 above)
Comment (kettlebell) [reply gf2_192/src/gf2_192poly.rs:172]: Sure.

File: gf2_192/src/lib.rs (hunk 5 above)
Comment (kettlebell) [reply gf2_192/src/lib.rs]: Sure.

File: gf2_192/src/gf2_192.rs [hunk 6]
```diff
@@ -0,0 +1,1025 @@
...
+        let mut a2muls: [i64; 16] = Default::default();
+
+        // a0muls[0], a1muls[0] and a2muls[0] are already correctly initialized to 0
+
+        a0muls[1] = a.word[0];
+        a1muls[1] = a.word[1];
+        a2muls[1] = a.word[2];
+
+        // a*x, a*x^2, a*x^3
+        for i in 2..=8 {
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply gf2_192/src/gf2_192.rs]: Thanks for spotting that! I corrected it, but it had no visible impact on the tests.

File: gf2_192/src/gf2_192poly.rs (hunk 2 above)
Comment (kettlebell) [reply gf2_192/src/gf2_192poly.rs:69]: I've added some more explanations; it covers everything from `sigmastate-interpreter`. Please let me know if it needs more.

File: ergotree-interpreter/src/sigma_protocol/gf2_192.rs [hunk 7]
```diff
@@ -1,19 +1,18 @@
-// TODO: remove after all todo! are implemented
-#![allow(clippy::todo)]
-
-use super::challenge::Challenge;
+use super::{challenge::Challenge, fiat_shamir::FiatShamirHash, SOUNDNESS_BYTES};
 
 #[derive(PartialEq, Debug, Clone)]
-pub(crate) struct Gf2_192 {}
+pub(crate) struct Gf2_192(pub(crate) gf2_192::gf2_192::Gf2_192);
```
[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-interpreter/src/sigma_protocol/gf2_192.rs:4]: We can ditch them, but I'll need to make a change to the interface for `GF2_192Poly`. The current wrapper for `Gf2_192Poly` takes care of u8 to i8 conversion (Reyzin's code works entirely with i8 due to Java/Scala having no unsigned ints). If we do away with it, I'll need to change/add-to the interface for `gf2_192::gf2_192poly::Gf2_192Poly`. Let me know which way you'd prefer.

[the remark this answers is not in the document]
Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/gf2_192.rs:4]: Yes, please move u8 to i8 conversion to `gf2_192` crate.

[the remark this answers is not in the document]
Comment (kettlebell) [reply ergotree-interpreter/src/sigma_protocol/gf2_192.rs]: Done.

File: gf2_192/src/gf2_192poly.rs (hunk 1 above)
Comment (kushti) [reply gf2_192/src/gf2_192poly.rs:219]: `new byte[0]` means empty array in Java.

thus `GF2_192_Poly.interpolate(new byte[0], new GF2_192[0], new GF2_192(0))` means that the only point among arguments which is defined is 0, and value is 0 

similarly, `GF2_192_Poly.interpolate(new byte[0], new GF2_192[0], val17)` means that f(0) = 17 and that is the only constraint 

so tests are correct

**Review by kushti — APPROVED** [review]

File: gf2_192/src/gf2_192poly.rs (hunk 1 above)
Comment (kettlebell) [reply gf2_192/src/gf2_192poly.rs:219]: > so tests are correct

Thank you for the explanation and for reviewing the PR!

**Review by greenhat — APPROVED** [review]

**Pull request merged**